### DISCUSSION TITLE
Install msixbundle Package Without need to internet

### DISCUSSION THREAD

**Discussion by ghost1372** [microsoft/WindowsAppSDK#1182] [Q&A] [answered]
I created a project as a test, after creating msixbundle file, I realized that when we install msixbundle, the desired framework (Project Reunion 0.8.0) will be installed through the store first. So if we want to install our program on an offline system, we will definitely have a problem.
Is there a solution to this problem?

**Comment by btueffers** [comment, marked as answer]
@DrusTheAxe Is there a way for apps to be installed offline?

**Reply by DrusTheAxe** [reply]
The [Add-AppxProvisionedPackage](https://docs.microsoft.com/en-us/powershell/module/dism/add-appxprovisionedpackage?view=windowsserver2019-ps) cmdlet lets you provision packages for online *and* offline systems. See [Example 2: Add an app package to operating system image](https://docs.microsoft.com/en-us/powershell/module/dism/add-appxprovisionedpackage?view=windowsserver2019-ps#example-2--add-an-app-package-an-operating-system-image) for more details :-)

The WinAppSDK installer doesn't install to an offline image (but it's an interesting idea) so you'll need to get the WinAppSDK MSIX packages and install them via Add-AppxProvisionedPackage along with your package(s).

@zaryaf got link to the directions/recipe how folks can directly install WinAppSDK's MSIXs (w/o using the installer)?

**Reply by DrusTheAxe** [reply]
[Installer: Add --offline support #1184](https://github.com/microsoft/WindowsAppSDK/issues/1184)

**Reply by riverar** [reply]
@ghost1372 Are you asking how to install all the dependencies into an offline image or simply cache all the files needed for install and disconnect from the Internet?

**Reply by zaryaf** [reply]
> @zaryaf got link to the directions/recipe how folks can directly install WinAppSDK's MSIXs (w/o using the installer)?

For an example that demonstrates how your setup program can install the MSIX packages directly without the installer, please take a look at the following: https://github.com/microsoft/WindowsAppSDK/blob/main/installer/dev/install.cpp

We also link to this in our developers docs and are planning to create a standalone sample in the near future.

**Reply by ghost1372** [reply]
> @ghost1372 Are you asking how to install all the dependencies into an offline image or simply cache all the files needed for install and disconnect from the Internet?

Hi @riverar  Yes we need offline image with all the dependencies. (All dependencies should be in the msixbundle file)
Thank you @DrusTheAxe I have another question, Is the installation of these dependencies just for WinUI 3 Packaged Project?
What about WinUI 3 UnPackaged Project? Do we not need to install dependencies?

**Reply by riverar** [reply]
@ghost1372 I'm still not sure if you're asking for the files ([they're in the nuget package](https://www.nuget.org/packages/Microsoft.ProjectReunion/)) or for instructions how to provision an offline windows image. ([You can provision appx/msix packages with DISM](https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/dism-app-package--appx-or-appxbundle--servicing-command-line-options).) Does that help?

**Reply by ghost1372** [reply]
@riverar When I transferred my WinUI 3 Packaged App project (msixbundle file) to another system, I noticed that first the Project Reunion app from the store is installed on the system (as a framework) and then the program itself is installed.
I do not have enough information about the contents of the nuget package or store app,
I expect the project I created (WinUI 3 Packaged App) to be installed on the user's system without the need for internet.

**Reply by DrusTheAxe** [reply]
What does "transferred" mean? Did you copy your .msixbundle to another machine? Did you do something to install it on this other system?

When you develop with VS and hit F5 everything should install and run without need for a network connection (let alone talking to the Store). There's a known issue where that doesn't happen in a certain circumstance (involving .NET 5 I think; I forget the detail other than it's on the punchlist to address for 1.0).

**Reply by ghost1372** [reply]
yes copy to another machine

**Reply by ghost1372** [reply]
Look at this gif to the end
First the framework is installed, then the program, the framework is installed through the store 
![h0](https://user-images.githubusercontent.com/9213496/129768811-4c25aef7-5117-4d4a-a430-8d94f97c602a.gif)

**Reply by zaryaf** [reply]
@ghost1372 Thank you for the gif, that helps a lot. Will your app be distributed to end users through the Store or do you have other distribution channels? If you deploy through the Store, the Store will handle the Framework package installation. If deployed outside of the Store using say an MSI or setup program, then you can integrate in our installer (which handles the scenario where you are disconnected from the internet). Alternatively, you can also directly carry the MSIX packages, which your setup program can install directly. So would appreciate a bit more information about your app, your users, and distribution channel.

**Reply by ghost1372** [reply]
Thanks for your explanation, my program will reach the end user outside the store Via msix file

**Reply by zaryaf** [reply]
@ghost1372 ok great. Can you please confirm if in your gif you were disconnected to the internet? Have you tested the installation when being offline - what happens?

**Reply by ghost1372** [reply]
@zaryaf  At the top gif, the internet was connected and msix installed without any problems. But in this new gif, the Internet is not connected (offline) and the msix is not installed and get following error

`App Installer failed to install package dependencies. Ask the developer for Microsoft.VCLibs.140.00 package. `

![003](https://user-images.githubusercontent.com/9213496/130049624-ea8d5b5a-94fc-451f-a741-ab381615154a.gif)

**Reply by zaryaf** [reply]
@ghost1372 . Your app was able to install packages that were already registered, but failed to install the non-registered framework package (VC Libs in this case) because for that, you would need to be connected to the network so the Store can grab it. Or alternatively, your app needs to be have opted in for offline licensing. https://docs.microsoft.com/en-us/microsoft-store/distribute-offline-apps
